### PR TITLE
profile elem accesses and emit fastpaths for cache types we've seen

### PR DESCRIPTION
**Pull request by MikeHolman**
Perf runs have shown a few scattered improvements of 2-4% across different Speedometer 2 tests (overall improves by ~0.8%). Also looks like it improves ARES-6 ML and Air tests by about 5% each.

### CODE REVIEW

File: lib/Runtime/Library/JavascriptArray.h [hunk 1]
```diff
@@ -1137,6 +1137,8 @@ namespace Js
         uint32 GetNextIndex(uint32 index) const;
         BOOL DirectGetItemAt(uint32 index, int* outVal);
 
+        virtual PropertyQueryFlags GetPropertyQuery(Var originalInstance, PropertyId propertyId, Var* value, PropertyValueInfo* info, ScriptContext* requestContext) override;
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptArray.h]: Looks like we also need `#if ENABLE_COPYONACCESS_ARRAY` for these method definitions.

Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptArray.h]: Do we need one of these for HasPropertyQuery too?

File: lib/Runtime/Language/ProfilingHelpers.cpp [hunk 2]
```diff
@@ -337,14 +353,34 @@ namespace Js
             }
         } while(false);
 
-        JavascriptOperators::OP_SetElementI(base, varIndex, value, scriptContext, flags);
+        RecyclableObject* cacheOwner;
+        PropertyRecordUsageCache* propertyRecordUsageCache;
+        TypeId instanceType = JavascriptOperators::GetTypeId(base);
+        bool isTypedArray = (instanceType >= TypeIds_Int8Array && instanceType <= TypeIds_Float64Array);
```
Comment (sethbrenith) [inline lib/Runtime/Language/ProfilingHelpers.cpp]: Why is this condition different than the one in ValueType::IsTypedArray? And why do we care here whether it's a typed array?

File: lib/Runtime/Language/JavascriptOperators.cpp [hunk 3]
```diff
@@ -3370,303 +3410,310 @@ namespace Js
...
+
+        if (JavascriptNumber::Is_NoTaggedIntCheck(index))
         {
             uint32 uint32Index = JavascriptConversion::ToUInt32(index, scriptContext);
 
             if ((double)uint32Index == JavascriptNumber::GetValue(index) && !TaggedInt::IsOverflow(uint32Index))
             {
                 index = TaggedInt::ToVarUnchecked(uint32Index);
-                goto TaggedIntIndex;
+                return GetElementIIntIndex(instance, index, scriptContext);
```
Comment (sethbrenith) [inline lib/Runtime/Language/JavascriptOperators.cpp]: Thanks, this feels much cleaner

File: lib/Runtime/Language/ProfilingHelpers.cpp (hunk 2 above)
Comment (MikeHolman) [reply lib/Runtime/Language/ProfilingHelpers.cpp]: I'm not sure why the conditions are different, but typed arrays have different semantics than other objects for OP_SetElementI.

File: lib/Runtime/Library/JavascriptArray.h (hunk 1 above)
Comment (MikeHolman) [reply lib/Runtime/Library/JavascriptArray.h]: I don't think so, but maybe need one fore GetItemQuery (the other path that GetElementI can go to).

**Comment by sethbrenith** [comment]
> // Currently only checks the local cache (not proto, accessor, or missing); we could consider generating extra checks

update comment

---
Refers to: lib/Backend/Lower.cpp:15698 in 181d889. [](commit_id = 181d889eb78fee61375aeb1cd2d8b89d7027133a, deletion_comment = False)

File: lib/Runtime/Library/JavascriptArray.h (hunk 1 above)
Comment (MikeHolman) [reply lib/Runtime/Library/JavascriptArray.h]: Actually nvm, NativeIntArray already has a GetItemQuery which does the check. Also upon consideration, I'm going to remove these overloads and just add checks to `JavascriptArray::GetPropertyQuery`. It's not a hot path and seems better not to be changing the vtable like this.

File: lib/Backend/Lower.cpp [hunk 4]
```diff
@@ -15682,52 +15722,227 @@ Lowerer::GenerateLookUpInIndexCache(IR::Instr * instrInsert, IR::RegOpnd *indexO
...
+                slotIndexLoadedLabel,
+                labelHelper,
+                &nextLabel,
+                &branchToPatch,
+                &taggedTypeOpnd);
+        }
     }
-    InsertBranch(Js::OpCode::Br, slotArrayLoadedLabel, instrInsert);
+    Assert(branchToPatch);
+    Assert(nextLabel);
```
Comment (sethbrenith) [inline lib/Backend/Lower.cpp]: While we're asserting things, perhaps we could assert that only one branch leads to nextLabel and it is branchToPatch

File: lib/Backend/Lower.cpp [hunk 5]
```diff
@@ -16895,7 +17111,10 @@ Lowerer::GenerateFastLdElemI(IR::Instr *& ldElem, bool *instrIsInHelperBlockRef)
             labelBailOut = IR::LabelInstr::New(Js::OpCode::Label, this->m_func, true);
             labelCantUseArray = labelBailOut;
         }
-
+        if (ldElem->IsProfiledInstr())
+        {
+            flags = ldElem->AsProfiledInstr()->u.ldElemInfo->flags;
```
Comment (sethbrenith) [inline lib/Backend/Lower.cpp]: Should we do the same in GenerateFastStElemI?

**Pull request merged**